src/Middleware/HttpCache.php

```diff
@@ -0,0 +1,197 @@
+<?php
+namespace Yiisoft\Yii\Web\Middleware;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Router\Method;
+
+/**
+ * HttpCache implements client-side caching by utilizing the `Last-Modified` and `ETag` HTTP headers.
+ */
+final class HttpCache implements MiddlewareInterface
+{
+    private const DEFAULT_HEADER = 'public, max-age=3600';
+
+    /**
+     * @var callable a PHP callback that returns the UNIX timestamp of the last modification time.
+     * The callback's signature should be:
+     *
+     * ```php
+     * function ($request, $params)
+     * ```
+     *
+     * where `$request` is the [[ServerRequestInterface]] object that this filter is currently handling;
+     * `$params` takes the value of [[params]]. The callback should return a UNIX timestamp.
+     *
+     * @see http://tools.ietf.org/html/rfc7232#section-2.2
+     */
+    private $lastModified;
+
+    /**
+     * @var callable a PHP callback that generates the ETag seed string.
+     * The callback's signature should be:
+     *
+     * ```php
+     * function ($request, $params)
+     * ```
+     *
+     * where `$request` is the [[ServerRequestInterface]] object that this filter is currently handling;
+     * `$params` takes the value of [[params]]. The callback should return a string serving
+     * as the seed for generating an ETag.
+     */
+    private $etagSeed;
+
+    /**
+     * @var bool whether to generate weak ETags.
+     *
+     * Weak ETags should be used if the content should be considered semantically equivalent, but not byte-equal.
+     *
+     * @see http://tools.ietf.org/html/rfc7232#section-2.3
+     */
+    private bool $weakEtag = false;
+
+    /**
+     * @var mixed additional parameters that should be passed to the [[lastModified]] and [[etagSeed]] callbacks.
+     */
+    private $params;
+
+    /**
+     * @var string the value of the `Cache-Control` HTTP header. If null, the header will not be sent.
+     * @see http://tools.ietf.org/html/rfc2616#section-14.9
+     */
+    private ?string $cacheControlHeader = self::DEFAULT_HEADER;
```

Comment:
Does it make sense to use this middleware without a header?

Reply:
Yes. See https://tools.ietf.org/html/rfc7232#section-2.3